torch/utils/_pytree.py

```diff
@@ -77,12 +81,28 @@ def _odict_flatten(d: 'OrderedDict[Any, Any]') -> Tuple[List[Any], Context]:
 def _odict_unflatten(values: List[Any], context: Context) -> 'OrderedDict[Any, Any]':
     return OrderedDict((key, value) for key, value in zip(context, values))
 
+def _dataclass_flatten(d: Any) -> Tuple[List[Any], Context]:
+    if not is_dataclass(d):
+        raise RuntimeError(
+            "_dataclass_flatten expects dataclass instance as input, but "
+            f"dataclasses.is_dataclass returns False. Got input type {type(d)}."
+        )
+
+    values, names = map(list, zip(*[[getattr(d, f.name), f.name] for f in fields(d)]))
+    return values, (type(d), names)
+
+def _dataclass_unflatten(values: List[Any], context: Context) -> Any:
+    clazz = context[0]
+    names = context[1]
+    return clazz(**{n: v for n, v in zip(names, values)})
+
 
 _register_pytree_node(dict, _dict_flatten, _dict_unflatten)
 _register_pytree_node(list, _list_flatten, _list_unflatten)
 _register_pytree_node(tuple, _tuple_flatten, _tuple_unflatten)
 _register_pytree_node(namedtuple, _namedtuple_flatten, _namedtuple_unflatten)
 _register_pytree_node(OrderedDict, _odict_flatten, _odict_unflatten)
+_register_pytree_node(dataclass, _dataclass_flatten, _dataclass_unflatten)
```

Comment:
Instead of registering all subclasses of `dataclass` as node type. Why not use `namedtuple`? Or register the target type manually. Having `EXPLICIT_LEAVES` is a bit weird to me.

Reply:
>  Why not use namedtuple?

It's not our choice. Users decide what their models return, and we have seem many users use dataclass as the return type. 

> Or register the target type manually. 

We are doing this now for some of the existing applications, but it won't scale and requires user code to rely on our internal APIs. Initially we wanted to do it within DDP, but the problem is that it won't work for nested dataclasses unless we reimplement pytree separately. Take the `OuterDataclass` as an example. DDP can register flatten/unflatten functions during runtime, after getting an `OuterDataclass` instance. However, its attribute b is a dict that contains another dataclass. We will need to recursively find all dataclass instances in the return value and register for all of them. It's doable, but it's a bit weird to have two pytree-like implementations in utils and DDP separately. Let us know if there is any other better solution to approach this problem. Thanks!

Reply:
> We will need to recursively find all dataclass instances in the return value and register for all of them.

I think you could use `metaclass` or `__init_subclass__` to register your dataclass automaticly. You can implement a `__init_subclass__` method in your base class, which calls `register_pytree_node(cls, ...)`. Then all its subclasses will automatically registered. And you can also leave other dataclass type as leaf rather than non-leaf node.

Reply:
This seems like it would require modifications to the user code (user's dataclass), which is likely a non-starter as we'd like to enable this change with no user code modifications needed.

Reply:
> This seems like it would require modifications to the user code (user's dataclass), which is likely a non-starter as we'd like to enable this change with no user code modifications needed.

Automatically registering all `dataclass` as node type is a breaking change. Users always need to update their code.

------

Before we make this change. There are still some issues that need to resolve.

### 1. Special type handling
 
You are assuming all users want `dataclass` to be a container, which may not be true. Someone may want dataclass to be an opaque object and cannot be flattened. There are already some discussions in https://github.com/goolge/jax. They are facing a similar issue:

- google/jax#2588
- https://github.com/google/jax/issues/2371#issuecomment-640055295

In this PR, a new registry of `EXPLICIT_LEAVES` is added. Having two register functions is a bit weird to me. It would be complex to understand and hard to document. 

### 2. Should a dataclass be considered a container (or specifically, a collection)?
 
By default, dataclass is not `Iterable`, it does not support `__iter__`, `__getitem__`, `__len__`, and `__contains__`. While `namedtuple` is collection class, it is a subclass of `tuple`. I think `dataclass` is more like the concept of `struct` in C/C++ rather than a collection type.

You could look into the definition of `collections.abc.Collection`:

```python
class Iterable(metaclass=ABCMeta):
    __slots__ = ()

    @abstractmethod
    def __iter__(self):
        while False:
            yield None

    @classmethod
    def __subclasshook__(cls, C):
        if cls is Iterable:
            return _check_methods(C, "__iter__")
        return NotImplemented

class Sized(metaclass=ABCMeta):
    __slots__ = ()

    @abstractmethod
    def __len__(self):
        return 0

    @classmethod
    def __subclasshook__(cls, C):
        if cls is Sized:
            return _check_methods(C, "__len__")
        return NotImplemented

class Container(metaclass=ABCMeta):
    __slots__ = ()

    @abstractmethod
    def __contains__(self, x):
        return False

    @classmethod
    def __subclasshook__(cls, C):
        if cls is Container:
            return _check_methods(C, "__contains__")
        return NotImplemented

class Collection(Sized, Iterable, Container):
    __slots__ = ()

    @classmethod
    def __subclasshook__(cls, C):
        if cls is Collection:
            return _check_methods(C,  "__len__", "__iter__", "__contains__")
        return NotImplemented
```

A collection should be a sized iterable container. None of the three conditions is fit to `dataclass`.

- https://github.com/google/jax/issues/2371#issuecomment-633680461

### 3. Class variables handling

Users can create a dataclass with both instance fields and class fields:

```python
from dataclasses import dataclass
from typing import ClassVar

@dataclass
class Foo:
    instance_field: int
    class_field: ClassVar[int]
```

Would a class variable be auxiliary data or a child node?

### 4. Attributes defined in `__post_init__`

Users can define custom logic by defining a `__post_init__` method. Users can add or delete some attributes. It's difficult to handle all these use cases.